docs(thanos): add more context and quick start guide (#477)
cloudoperators/greenhouse-extensions: thanos/README.md

title: Thanos
---

Learn more about the **Thanos** Plugin. Use it to enable extended metrics retention and querying across Prometheus servers and Greenhouse clusters. 

The main terminologies used in this document can be found in [core-concepts](https://cloudoperators.github.io/greenhouse/docs/getting-started/core-concepts).

## Overview

> Observability is often required for operation and automation of service offerings. To get the insights provided by an application and the container runtime environment, you need telemetry data in the form of _metrics_ or _logs_ sent to backends such as _Prometheus_ or _OpenSearch_. With the **Thanos** plugin you can scale the _metric_ part of the observability stack horizontally by hooking into the Prometheus provided by the **kube-monitoring** plugin and optionally save your data in the object storage.

Reviewer: ```suggestion
Observability is required for monitoring the lifecycle of service offerings. To get the insights provided by an application and the container runtime environment, you need telemetry data in the form of _metrics_ or _logs_ which are made accessible through backends such as _Prometheus_ or _OpenSearch_.
With the **Thanos** plugin you can enhance the _metric_ part of the observability stack to scale horizontally and leverage object storage for a long retention setup.
```

What explicitly do you mean with `container runtime environment`?

To me this is/was somehow hard to read and grasp. I rewrote it, not sure if this is understandable - or even needed. The following section (`In general, Thanos is...`)is more to the point, while this is too generic. I would wipe it entirely. Similar wording could go in an overall metrics architecture document.

Furthermore, I don't see why logs w/ need to be mentioned here, this also would rather be part of an overarching document.

Author: Kubernetes. 🤣 
> What explicitly do you mean with container runtime environment?

You're right, I admit. We don't have space for such a general document. That's why it's included in every plugin at the moment. We can also remove it and I will find a place for it.

> To me this is/was somehow hard to read and grasp. I rewrote it, not sure if this is understandable - or even needed. The following section (In general, Thanos is...)is more to the point, while this is too generic. I would wipe it entirely. Similar wording could go in an overall metrics architecture document.
> Furthermore, I don't see why logs w/ need to be mentioned here, this also would rather be part of an overarching document.

> In general, Thanos is a set of components that can be used to extend the storage and retrieval of metrics in Prometheus. It allows you to store metrics in a remote object store and query them across multiple Prometheus servers and Greenhouse clusters. This Plugin is intended to provide a set of pre-configured Thanos components that enable a proven composition. At the core, a set of Thanos components is installed that adds long-term storage capability to a single **kube-monitoring** Plugin and makes both current and historical data available again via one Thanos Query component.

Reviewer: I am now asking myself, if this is sort of a repetition of the very first sentence

> Learn more about the **Thanos** Plugin. Use it to enable extended metrics retention and querying across Prometheus servers and Greenhouse clusters. 

This part could easily go there:
> This Plugin is intended to provide a set of pre-configured Thanos components that enable a proven composition. At the core, a set of Thanos components is installed that adds long-term storage capability to a single **kube-monitoring** Plugin and makes both current and historical data available again via one Thanos Query component.

We could however add some prose on how the components are acting together here?

Author: You are right. We can improve.
> I am now asking myself, if this is sort of a repetition of the very first sentence

Good idea. I'm open for a contribution from your side. 
> We could however add some prose on how the components are acting together here?

This plugin deploys the following Thanos components:

* [Thanos Query](https://thanos.io/tip/components/query.md/)
* [Thanos Store](https://thanos.io/tip/components/store.md/)
* [Thanos Compactor](https://thanos.io/tip/components/compact.md/)

Planned components:
* [Thanos Ruler](https://thanos.io/tip/components/rule.md/)
* [Thanos Query Frontend](https://thanos.io/tip/components/query.md/)

This Plugin does **not** deploy the following components:
* [Thanos Sidecar](https://thanos.io/tip/components/sidecar.md/)
This component is installed in the [kube-monitoring](https://github.com/cloudoperators/greenhouse-extensions/tree/main/kube-monitoring) plugin.

## Note

It is not meant to be a comprehensive package that covers all scenarios. If you are an expert, feel free to configure the Plugin according to your needs.

Contribution is highly appreciated. If you discover bugs or want to add functionality to the plugin, then pull requests are always welcome.

## Quick start

This guide provides a quick and straightforward way to use **Thanos** as a Greenhouse Plugin on your Kubernetes cluster. The guide is meant to build the following setup:

> ![Thanos Architecture](./img/Thanos-setup-example.png)

Reviewer: Cool pic.

I would change the wording of `range queries` to `long term queries`.

A range query would also be a query for the last `[1h]` and this would be resolved by the sidecar. We can also be more precise and note `metrics older than 2h` or similar.

**Prerequisites**

> - A running and Greenhouse-onboarded Kubernetes cluster

Reviewer: Do we have some link here?

The main terminologies used in this document can be found in [core-concepts](https://cloudoperators.github.io/greenhouse/docs/getting-started/core-concepts).

## Overview

Thanos is a set of components that can be used to extend the storage and retrieval of metrics in Prometheus. It allows you to store metrics in a remote object store and query them across multiple Prometheus servers and Greenhouse clusters. This Plugin is intended to provide a set of pre-configured Thanos components that enable a proven composition. At the core, a set of Thanos components is installed that adds long-term storage capability to a single **kube-monitoring** Plugin and makes both current and historical data available again via one Thanos Query component.

![Thanos Architecture](./img/Thanos-setup-example.png)

The **Thanos Sidecar** is a component that is deployed as a container together with a Prometheus instance. This allows Thanos to optionally upload metrics to the object store and Thanos Query to access Prometheus data via a common, efficient StoreAPI.

> The **Thanos Compact** component applies the Prometheus 2.0 Storage Engine compaction process to block data uploaded to the object store. The Compactor is also responsible for applying the configured retention and downsampling of the data. 

Reviewer: ```suggestion
The **Thanos Compact** component applies the Prometheus 2.0 Storage Engine compaction process to data uploaded to the object store. The Compactor is also responsible for applying the configured retention and downsampling of the data. 
```

I find block data a little misleading, as it reads like the verb "to block" data.

> The **Thanos Store** also implements the Store API but for historical data in an object store. It acts primarily as an API gateway and therefore does not require large amounts of local storage space.

Reviewer: ```suggestion
The **Thanos Store** also implements the StoreAPI and serves the historical data from an object store. It acts primarily as an API gateway and has no persistence itself.
```

> **Thanos Query** implements the Prometheus HTTP v1 API for querying data in a Thanos cluster via PromQL. In short, it collects the data needed to evaluate the query from the underlying StoreAPIs, evaluates the query and returns the result.

Reviewer: ```suggestion
**Thanos Query** implements the Prometheus HTTP v1 API for querying data in a Thanos cluster via PromQL. In short, it collects the data needed to evaluate the query from the connected StoreAPIs, evaluates the query and returns the result.
```

This plugin deploys the following Thanos components:

* [Thanos Query](https://thanos.io/tip/components/query.md/)
* [Thanos Store](https://thanos.io/tip/components/store.md/)
* [Thanos Compactor](https://thanos.io/tip/components/compact.md/)

Planned components:
* [Thanos Ruler](https://thanos.io/tip/components/rule.md/)
* [Thanos Query Frontend](https://thanos.io/tip/components/query.md/)

This Plugin does **not** deploy the following components:
* [Thanos Sidecar](https://thanos.io/tip/components/sidecar.md/)
This component is installed in the [kube-monitoring](https://github.com/cloudoperators/greenhouse-extensions/tree/main/kube-monitoring) plugin.

## Disclaimer 

It is not meant to be a comprehensive package that covers all scenarios. If you are an expert, feel free to configure the Plugin according to your needs.

Contribution is highly appreciated. If you discover bugs or want to add functionality to the plugin, then pull requests are always welcome.

## Quick start

> This guide provides a quick and straightforward way to use **Thanos** as a Greenhouse Plugin on your Kubernetes cluster. The guide is meant to build the following setup:

Reviewer: ```suggestion
This guide provides a quick and straightforward way to use **Thanos** as a Greenhouse Plugin on your Kubernetes cluster. The guide is meant to build the following setup.
```

kubectl create secret generic $THANOS_PLUGIN_NAME-metrics-objectstore --from-file=thanos.yaml=/path/to/your/file
```

### kube-monitoring plugin enablement 

> (usually with [Prometheus Operator](https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.ThanosSpec)).

Reviewer: Would move this at the end of this section and rename it:

```Values used here are described in the [Prometheus Operator Spec](https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.ThanosSpec).```

### Standalone Query

![Standalone Query](./img/Thanos-standalone-query.png)

> In case you want to achive a Setup like the above and have a Thanos Query to run with multiple Stores, you can set it to `standalone` and add your own store list. Setup your Plugin like this:

Reviewer: ```suggestion
In case you want to achieve a setup like above and have an overarching Thanos Query to run with multiple Stores, you can set it to `standalone` and add your own store list. Setup your Plugin like this:
```
